### PR TITLE
Adding functionality to cvxpy/tests: verifying stationarity of the Lagrangians as part of KKT conditions

### PR DESCRIPTION
**Pull request by aryamanjeendgar**
## Description
This PR adds a new method, `check_stationarity_lagrangian`, which:
1. Tries to construct the Lagrangian corresponding to a given problem
2. Check if it is stationary by using the existing `.grad` facility.
3. Can also handle the case of the user passing in implicit constraints via flags (such as `diag, PSD, NSD, nonpos` etc. by checking dual cone membership of the computed gradient instead of stationarity)

I have added in tests for this new method which live in `cvxpy/tests/test_KKT.py` --- where all of the standard test classes from `SolverTestHelpers` make an appearance --- the KKT conditions are verified in full for all of the tests (with the exception of MI-programs)  in the `solver_test_helpers.py` file.

I haven't added documentation for this method yet since it just lives within the `SolverTestHelper` class as of now --- we can probably write a full class for verifying the KKT conditions after proper support for verifying dual feasibility has been added (via the introduction of the `dual` method on `Constraint` objects).

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/tests/test_KKT.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+    def test_lp_5(self):
+        # LP with redundant constraints
+        sth = STH.lp_5()
+        sth.solve(solver='ECOS')
+        sth.verify_objective(places=4)
+        sth.check_primal_feasibility(places=4)
+        sth.check_complementarity(places=4)
+        sth.check_dual_domains(places=4)
+        sth.check_stationary_lagrangian(places=2)
```
Comment (rileyjmurray) [inline cvxpy/tests/test_KKT.py:36]: You don't need to use ``verify_objective``, or ``verify_primal_values``, or ``verify_dual_values`` here. Just check the KKT conditions.

File: cvxpy/tests/test_KKT.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        sth.solve(solver='ECOS')
+        sth.verify_primal_values(places=4)
+        sth.verify_objective(places=4)
+        sth.check_complementarity(places=4)
+        sth.verify_dual_values(places=4)
+        sth.check_stationary_lagrangian(places=4)
+        return sth
+
+
+class StandardTestSOCPs(BaseTest):
```
Comment (rileyjmurray) [inline cvxpy/tests/test_KKT.py:52]: Update these tests to only check the KKT conditions.

File: cvxpy/tests/test_KKT.py
Comment (rileyjmurray) [inline cvxpy/tests/test_KKT.py:1]: Naming conventions of classes: test classes start with ``Test...``. The convention you're using here is like for classes in ``solver_test_helpers.py``, which is not actually dispatched by a test runner.

File: cvxpy/tests/test_KKT.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+
+class StandardTestFlags(BaseTest):
+    """
+    A class with tests for testing out the incorporation of
+    implicit constraints within `check_stationarity_lagrangian`
+    """
+
+    @staticmethod
+    def tf_1() -> STH.SolverTestHelper:
```
Comment (rileyjmurray) [inline cvxpy/tests/test_KKT.py:183]: What's the ``tf`` stand for here?

File: cvxpy/tests/test_KKT.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        sth.solve(solver='SCS')
+        sth.verify_objective(places=2)
+        sth.check_primal_feasibility(places=2)
+        sth.check_complementarity(places=2)
+        sth.check_dual_domains(places=2)
+        sth.check_stationary_lagrangian(places=2)
+        return sth
+
+    def test_sdp_2(self):
+        # places is set to 3 rather than 4, because analytic solution isn't known.
```
Comment (rileyjmurray) [inline cvxpy/tests/test_KKT.py:128]: This is actually a great example where we don't need to know the analytic solution. Just check the KKT conditions to the usual level of accuracy.

File: cvxpy/tests/test_KKT.py
Comment (rileyjmurray) [reply cvxpy/tests/test_KKT.py:1]: Separately: there are no tests for exponential cone constraints here.

**Review by rileyjmurray — CHANGES_REQUESTED** [review]
See my individual comments.

**Comment by aryamanjeendgar** [comment]
@rileyjmurray I think I should've addressed your suggested changes with this commit, please take a look whenever you get time!

**Comment by rileyjmurray** [comment]
Failing tests are unrelated to this PR.

File: cvxpy/tests/test_KKT.py (hunk 3 above)
Comment (rileyjmurray) [reply cvxpy/tests/test_KKT.py]: The nature of the flag isn't obvious from the name itself. (Resolved with the latest naming convention I committed to your branch.)

**Review by rileyjmurray — APPROVED** [review]

**Pull request merged**